Add Chirpstack packages

stv0g opened the pull request:
## Description of changes

This PR adds the following packages from the [Chirpstack](https://www.chirpstack.io/) project:

- chirpstack-concentratord
- chirpstack-udp-forwarder
- chirpstack-mqtt-forwarder
- chirpstack-gateway-bridge
- chirpstack-rest-api
- chirpstack-fuota-server

> ChirpStack is an open-source LoRaWAN Network Server which can be used to setup LoRaWAN networks. ChirpStack provides a web-interface for the management of gateways, devices and tenants as well to setup data integrations with the major cloud providers, databases and services commonly used for handling device data. ChirpStack provides a gRPC based API that can be used to integrate or extend ChirpStack.

These are most of the packages of the Chirpstack project with exception of the main LoraWAN server itself. It is sufficient to run a Nix-powered LoraWAN gateway on an embedded device like a Raspberry Pi.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

AndersonTorres commented on pkgs/by-name/ch/chirpstack-udp-forwarder/package.nix:
```diff
@@ -0,0 +1,30 @@
+    owner = "chirpstack";
+    repo = "chirpstack-udp-forwarder";
+    rev = "81dbf333d7d0fca08e025a2ed801a7592d1cecb8";
+    sha256 = "sha256-JRwWD5TJCibTNRmn69lBPT+aJ/91y92Y7+hbJKxc3BE=";
```
use `hash` instead

stv0g replied:
Thanks, I am still a Nixpkgs newbie :D I addressed it in my latest push.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/3787

stv0g commented:
I've updated the branch:

- Rebase to `master`
- Reformatted with RFC166-style nixfmt
- Added `passthru.updateScript`
- Use Git tags for `fetchFromGitHub`'s `rev` attribute

drupol commented on pkgs/by-name/ch/chirpstack-concentratord/package.nix:
```diff
@@ -0,0 +1,130 @@
+      "libloragw"
+    ];
+
+    installPhase = ''
```
If you are overriding `configurePhase`, `buildPhase`, `checkPhase`, `installPhase` or any other phase, you should not forget about explicitly running pre and post-phase hooks like [their](https://github.com/NixOS/nixpkgs/blob/d71a03ad695407dd482ead32d3eddff50092a1c3/pkgs/stdenv/generic/setup.sh#L953) [original](https://github.com/NixOS/nixpkgs/blob/d71a03ad695407dd482ead32d3eddff50092a1c3/pkgs/stdenv/generic/setup.sh#L1008) [definitions](https://github.com/NixOS/nixpkgs/blob/d71a03ad695407dd482ead32d3eddff50092a1c3/pkgs/stdenv/generic/setup.sh#L1037) [do](https://github.com/NixOS/nixpkgs/blob/d71a03ad695407dd482ead32d3eddff50092a1c3/pkgs/stdenv/generic/setup.sh#L1078).

It is generally expected that an appropriate pre-phase hook (e.g. `preBuild`) hook will run at the beginning of a phase (e.g. `buildPhase`) and post-phase hook (e.g. `postBuild`) will run at the end. Hooks are normally ran as a part of a phase so if you override a phase as a whole, you will need to add `runHook hookName` calls (e.g. `runHook preBuild`) manually.

Having phases run pre/post-phase hooks is important because many setup hooks insert their own code into them – omitting a hook might therefore prevent some setup hooks required for proper functionality of a package from running. Additionally, hooks are often inserted by developers into the package expression and by users when overriding a package using `overrideAttrs`. Not running them can thus cause confusion why their code is not executed.

## Examples

### Before

```nix
  installPhase = ''
    your commands
  '';
```

### After

```nix
  installPhase = ''
    runHook preInstall

    your commands

    runHook postInstall
  '';
```

Source: https://github.com/jtojnar/nixpkgs-hammering/blob/main/explanations/missing-phase-hooks.md

stv0g replied:
Fixed in ea05670e75d4

drupol commented on pkgs/by-name/ch/chirpstack-concentratord/package.nix:
```diff
@@ -0,0 +1,130 @@
+    preBuildPhases = [ "buildDeps" ];
+
+    installPhase = ''
+      mkdir -p $out/{lib,include/libloragw-sx1302}
```
If you are overriding `configurePhase`, `buildPhase`, `checkPhase`, `installPhase` or any other phase, you should not forget about explicitly running pre and post-phase hooks like [their](https://github.com/NixOS/nixpkgs/blob/d71a03ad695407dd482ead32d3eddff50092a1c3/pkgs/stdenv/generic/setup.sh#L953) [original](https://github.com/NixOS/nixpkgs/blob/d71a03ad695407dd482ead32d3eddff50092a1c3/pkgs/stdenv/generic/setup.sh#L1008) [definitions](https://github.com/NixOS/nixpkgs/blob/d71a03ad695407dd482ead32d3eddff50092a1c3/pkgs/stdenv/generic/setup.sh#L1037) [do](https://github.com/NixOS/nixpkgs/blob/d71a03ad695407dd482ead32d3eddff50092a1c3/pkgs/stdenv/generic/setup.sh#L1078).

It is generally expected that an appropriate pre-phase hook (e.g. `preBuild`) hook will run at the beginning of a phase (e.g. `buildPhase`) and post-phase hook (e.g. `postBuild`) will run at the end. Hooks are normally ran as a part of a phase so if you override a phase as a whole, you will need to add `runHook hookName` calls (e.g. `runHook preBuild`) manually.

Having phases run pre/post-phase hooks is important because many setup hooks insert their own code into them – omitting a hook might therefore prevent some setup hooks required for proper functionality of a package from running. Additionally, hooks are often inserted by developers into the package expression and by users when overriding a package using `overrideAttrs`. Not running them can thus cause confusion why their code is not executed.

## Examples

### Before

```nix
  installPhase = ''
    your commands
  '';
```

### After

```nix
  installPhase = ''
    runHook preInstall

    your commands

    runHook postInstall
  '';
```

Source: https://github.com/jtojnar/nixpkgs-hammering/blob/main/explanations/missing-phase-hooks.md

stv0g replied:
Fixed in ea05670e75d4

drupol commented on pkgs/by-name/ch/chirpstack-fuota-server/package.nix:
```diff
@@ -0,0 +1,40 @@
+    homepage = "https://www.chirpstack.io/";
+    license = licenses.mit;
+    maintainers = with maintainers; [ stv0g ];
+    platforms = [
```
I think there's a shortcut for this.

stv0g replied:
Fixed in d1c5d605b072

SuperSandro2000 commented on pkgs/by-name/ch/chirpstack-concentratord/package.nix:
```diff
@@ -0,0 +1,131 @@
+
+    buildDeps = ''
+      make -C libtools
+    '';
```
```suggestion

    preBuild = ''
      make -C libtools
    '';

    installPhase = ''
      mkdir -p $out/{lib,include/libloragw-2g4}
      cp libloragw/libloragw.a $out/lib/libloragw-2g4.a
      cp libloragw/inc/* $out/include/libloragw-2g4
    '';
```

stv0g replied:
Done 👍🏻

SuperSandro2000 commented on pkgs/by-name/ch/chirpstack-concentratord/package.nix:
```diff
@@ -0,0 +1,131 @@
+
+    buildDeps = ''
+      make -C libtools
+    '';
```
```suggestion

    preBuild = ''
      make -C libtools
    '';

    installPhase = ''
      mkdir -p $out/{lib,include/libloragw-sx1302}
      cp libloragw/libloragw.a $out/lib/libloragw-sx1302.a
      cp libloragw/inc/* $out/include/libloragw-sx1302
      cp libtools/*.a $out/lib/
      cp libtools/inc/* $out/include/
    '';
```

stv0g replied:
Done 👍🏻

SuperSandro2000 commented on pkgs/by-name/ch/chirpstack-gateway-bridge/package.nix:
```diff
@@ -0,0 +1,33 @@
+    homepage = "https://www.chirpstack.io/";
+    license = licenses.mit;
+    maintainers = with maintainers; [ stv0g ];
+    platforms = [ "aarch64-linux" "x86_64-linux" "aarch64-darwin" "x86_64-darwin" ];
```
```suggestion
```
that is the default, we should not repeat it

stv0g replied:
Done 👍🏻

SuperSandro2000 commented on pkgs/by-name/ch/chirpstack-mqtt-forwarder/package.nix:
```diff
@@ -0,0 +1,39 @@
+    homepage = "https://www.chirpstack.io/";
+    license = licenses.mit;
+    maintainers = with maintainers; [ stv0g ];
+    platforms = [ "aarch64-linux" "x86_64-linux" "aarch64-darwin" "x86_64-darwin" ];
```
```suggestion
```

stv0g replied:
Done 👍🏻

SuperSandro2000 commented on pkgs/by-name/ch/chirpstack-rest-api/package.nix:
```diff
@@ -0,0 +1,33 @@
+    homepage = "https://www.chirpstack.io/";
+    license = licenses.mit;
+    maintainers = with maintainers; [ stv0g ];
+    platforms = [ "aarch64-linux" "x86_64-linux" "aarch64-darwin" "x86_64-darwin" ];
```
```suggestion
```

stv0g replied:
Done 👍🏻

SuperSandro2000 commented on pkgs/by-name/ch/chirpstack-udp-forwarder/package.nix:
```diff
@@ -0,0 +1,37 @@
+    homepage = "https://www.chirpstack.io/";
+    license = licenses.mit;
+    maintainers = with maintainers; [ stv0g ];
+    platforms = [ "aarch64-linux" "x86_64-linux" "aarch64-darwin" "x86_64-darwin" ];
```
```suggestion
```

SuperSandro2000 commented on pkgs/by-name/ch/chirpstack-fuota-server/package.nix:
```diff
@@ -0,0 +1,33 @@
+, fetchFromGitHub
+}: buildGoModule rec {
+  pname = "chirpstack-fuota-server";
+  name = pname;
```
```suggestion
```

SuperSandro2000 commented on pkgs/by-name/ch/chirpstack-fuota-server/package.nix:
```diff
@@ -0,0 +1,33 @@
+  ldflags = [
+    "-s"
+    "-w"
+    # "-X main.version=${version}"
```
```suggestion
```

SuperSandro2000 commented on pkgs/by-name/ch/chirpstack-fuota-server/package.nix:
```diff
@@ -0,0 +1,33 @@
+    homepage = "https://www.chirpstack.io/";
+    license = licenses.mit;
+    maintainers = with maintainers; [ stv0g ];
+    platforms = [ "aarch64-linux" "x86_64-linux" "aarch64-darwin" "x86_64-darwin" ];
```
```suggestion
```

stv0g replied:
Done 👍🏻

SuperSandro2000 commented on pkgs/by-name/ch/chirpstack-concentratord/package.nix:
```diff
@@ -1,8 +1,9 @@
-{ lib
```
Please squash this commit into the init commits

stv0g replied:
Done 👍🏻

SuperSandro2000 commented on pkgs/by-name/ch/chirpstack-concentratord/package.nix:
```diff
@@ -3,6 +3,7 @@
   stdenv,
```
same with this commit

stv0g replied:
Done 👍🏻

SuperSandro2000 commented on pkgs/by-name/ch/chirpstack-gateway-bridge/package.nix:
```diff
@@ -11,7 +11,7 @@ buildGoModule rec {
   src = fetchFromGitHub {
```
and this one

stv0g replied:
Done 👍🏻

SuperSandro2000 commented on pkgs/by-name/ch/chirpstack-concentratord/package.nix:
```diff
@@ -51,11 +51,15 @@ let
     preBuildPhases = [ "buildDeps" ];
```
and this commit

stv0g replied:
Done 👍🏻

stv0g committed:
chirpstack-concentratord: init at 4.4.2

stv0g committed:
chirpstack-fuota-server: init at unstable-2024-04-02

stv0g committed:
chirpstack-gateway-bridge: init at 4.0.11

stv0g committed:
chirpstack-udp-forwarder: init at 4.1.8

stv0g committed:
chirpstack-mqtt-forwarder: init at 4.3.1

stv0g committed:
chirpstack-rest-api: init at 4.9.0

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/4491

stv0g commented:
@SuperSandro2000 @AndersonTorres Do you maybe have some bandwidth to merge this PR? I have addressed your previous comments already. And pipelines are passing now :)

stv0g commented on pkgs/by-name/ch/chirpstack-fuota-server/package.nix:
```diff
@@ -0,0 +1,32 @@
+  fetchFromGitHub,
+}:
+buildGoModule rec {
+  name = "chirpstack-fuota-server";
```
There is no stable release yet for this component of the Chirpstack stack.

I've created a ticket upstream to release a new version in https://github.com/chirpstack/chirpstack-fuota-server/issues/17

PJungkamp replied:
You should use the current standard for unstable versions. Since there isn't any tagged release that would be `0-unstable-YYYY-MM-DD`. With the author date of the latest commit you're packaging.

See https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-naming

stv0g replied:
Thanks. I've addressed it in the latest version of the MR

PJungkamp commented on pkgs/by-name/ch/chirpstack-concentratord/package.nix:
```diff
@@ -0,0 +1,134 @@
+  protobuf,
+}:
+let
+  libloragw-sx1301 = stdenv.mkDerivation rec {
```
And if these were to be kept within this derivation, you should at least provide them in `passthru`, so people using `overrideAttrs` or similar have access to these libraries.

See https://nixos.org/manual/nixpkgs/stable/#var-stdenv-passthru

stv0g replied:
Done. I've packaged them as three individual libs.

pjungkamp reviewed (commented):
You might want to consider removing the `with lib;` statements from `meta` attributes.

stv0g committed:
libloragw-2g4: init at 1.1.0

stv0g committed:
libloragw-sx1301: init at 5.0.1r2

stv0g committed:
libloragw-sx1302: init at 2.1.0r7

stv0g commented:
> You might want to consider removing the `with lib;` statements from `meta` attributes.

Done in the latest version of the PR.

PJungkamp commented:
Does this currently build correctly on `aarch64-linux`? I just tried to cross-compile using `pkgsCross.aarch64-multiplatform` on an `x86_64-linux` machine but one of the libloragw libraries failed to build.

stv0g commented:
@PJungkamp no cross builds have not been tested. I will have a closer look at it.

stv0g commented:
I've split this into several smaller PRs, as those hopefully get merged more easily:

- https://github.com/NixOS/nixpkgs/pull/353921
- https://github.com/NixOS/nixpkgs/pull/353920
- https://github.com/NixOS/nixpkgs/pull/353919
- https://github.com/NixOS/nixpkgs/pull/353918
- https://github.com/NixOS/nixpkgs/pull/353917
- https://github.com/NixOS/nixpkgs/pull/353916
